// react-native-avoid-softinput/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  AvoidSoftInput: true,
  AvoidSoftInputView: true,
  useSoftInputAppliedOffsetChanged: true,
  useSoftInputHeightChanged: true,
  useSoftInputHidden: true,
  useSoftInputShown: true,
  useSoftInputState: true
};
Object.defineProperty(exports, "AvoidSoftInput", {
  enumerable: true,
  get: function () {
    return _AvoidSoftInput.AvoidSoftInput;
  }
});
Object.defineProperty(exports, "AvoidSoftInputView", {
  enumerable: true,
  get: function () {
    return _AvoidSoftInputView.AvoidSoftInputView;
  }
});
Object.defineProperty(exports, "useSoftInputAppliedOffsetChanged", {
  enumerable: true,
  get: function () {
    return _hooks.useSoftInputAppliedOffsetChanged;
  }
});
Object.defineProperty(exports, "useSoftInputHeightChanged", {
  enumerable: true,
  get: function () {
    return _hooks.useSoftInputHeightChanged;
  }
});
Object.defineProperty(exports, "useSoftInputHidden", {
  enumerable: true,
  get: function () {
    return _hooks.useSoftInputHidden;
  }
});
Object.defineProperty(exports, "useSoftInputShown", {
  enumerable: true,
  get: function () {
    return _hooks.useSoftInputShown;
  }
});
Object.defineProperty(exports, "useSoftInputState", {
  enumerable: true,
  get: function () {
    return _hooks.useSoftInputState;
  }
});
var _AvoidSoftInput = require("./AvoidSoftInput");
var _AvoidSoftInputView = require("./AvoidSoftInputView");
var _hooks = require("./hooks");
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
